docs/spec/identities.md

```diff
@@ -0,0 +1,438 @@
+---
+title: "Radicle Link -- Section: Identities"
+author:
+- Kim Altintop[^0]
+status: draft
+date: 2020-07-15
+
+numbersections: true
+link-citations: true
+bibliography: references.bib
+...
+[^0]: [kim@monadic.xyz](mailto:kim@monadic.xyz)
+
+# Identities
+
+In order to collaborate on repositories within a consensus-free network, we must
+be able to refer to them using a stable identifier. Note that this identifier is
+a **statement of intent**: a repository can be described as a collection of
+ever-moving leaves of a DAG whose root element is the empty object. Therefore,
+the _content_ of a repository is not enough to describe it -- while two views on
+the repository may share objects, they may diverge substantially otherwise. Both
+views may however state their intent to eventually converge to the same state.
+
+While in principle a random identifier with sufficient entropy would suffice for
+the purpose, this would put the burden of deciding which repository views are
+_legit_ (for some definition of legit) entirely on the user. Instead, our
+approach is to establish an ownership proof, tied to the network identity of a
+peer (`PeerId`), or a set of peers, such that repository views can be replicated
+according to the trust relationships between peers ("tracking").
+
+Our model is based on The Update Framework (TUF)[@tuf], conveived as a means of
+securely distributing software packages.
+
+## Data Model
+
+The identity of a repository is established using a document of the form:
+
+```rust
+struct Doc<T, D> {
+    replaces: Option<Revision>,
+    payload: T,
```

Comment:
I'd really like to see a storage model where the `payload: T,` property of the "document" is split out in a separated blob.

The rationale is that radicle link is not concerned with that payload.

Putting it inside the document blob forces at least a json-compatible format on it, but most of all it forces radicle link to deal with that payload when deserializing the document metadata that it cares about, when that payload should be totally ignored (and possibly opaque).

Putting it in a separated blob would allow the application layer to evolve the `T` type without changing the radicle link code.

Reply:
I disagree: the extensibility and compatibility requirements exist in any case -- every application that wishes to put extra metadata there MUST be able to deal with payloads it doesn't know about. Instead of requiring every application to deal with a serialisation scheme themselves, we can devise a parser which simply preserves unknown portions of the payload as untyped `Value`s, forcing applications to always treat things they expect as optional fields. I don't see an advantage of a separate file for this.

Reply:
I believe the idea is that `payload: T` _is the blob_ contents. The fact the `T` is generic means the application layer can evolve the `T` however they want because they determine the `Deserialize` impls. Radicle link shouldn't care what's stored in there as long as it can de/serialise it.

Reply:
> I don't see an advantage of a separate file for this.

Radicle link is only concerned with the `replaces` `delegations` properties from the `Doc`, which are tiny compared to `payload: T`.

The `replaces` and `delecations` properties are radicle link's business, their eventual evolution is radicle link's concern, but the `payload` is not so.

Splitting out the payload allows Radicle link to have:

* no need to parse it
* no need to even load it in memory
* no concern about its format at all, including
  - no need no need to play deserialization tricks to parse it in a free form way if its type is unknown, or
  - no need to know its type at all, or
  - no need to be updated every time Upstream evolves the `T` type if we implement deserialization in such a way that the payload structure must be known (which is the trivial way to implement it)

I would count these as advantages.

Reply:
I'm not sure your counter-proposal is clear to me. Can you expand on what splitting out the payload would look like?

Reply:
Let recap the two data structures in the current proposal and how they map to git contents (as I understood it):


The "document":
```
struct Doc<T, D> {
    /// Optional OID of the "parent" revision document
    replaces: Option<Revision>,
    /// Opaque application specific data
    payload: T,
    /// Roughly, the set of trusted keys (related to users in the case of projects)
    delegations: D,
}
```

The "identity"
```
struct Identity<T, D> {
    /// The OID of the identity revision commit
    id: ContentId,
    /// The OID of the identity revision document
    /// (pointed to by the tree)
    root: Revision,
    /// The OID of the tree referred by the identity commit
    revision: Revision,
    /// The document is referred to by the tree entry with name "root"
    /// (it also has OID "root")
    doc: Doc<T, D>,
    /// Commit trailers
    signatures: HashMap<PublicKey, Signature>,
}

```

So an identity is a commit pointing to a tree and containing signatures as commit trailers.

My simpler proposal would be to take the `replaces` and `delegations` properties away from the `Doc` and to put them:

1. either into another blob pointed to by the tree,
2. or inside other commit trailers or anyway some kind of encoding in the commit message text, in a proper canonical format.

Choice between 1 and 2 is an implementation detail, I'd have my preference but ultimately I care more about the separation than the actual encoding.

Going for 2 (so I don't have to define more tree entries and their naming) the identity encoding would be:

```
struct Identity<T, D> {
    /// The OID of the identity revision commit
    id: ContentId,
    /// Optional OID of the "parent" revision document, extracted from the commit message
    replaces: Option<Revision>,
    /// Roughly, the set of trusted keys (extracted from the commit message)
    delegations: D,
    /// The OID of the identity revision payload document
    /// (pointed to by the tree)
    root: Revision,
    /// The OID of the tree referred by the identity commit
    revision: Revision,
    /// The payload document is referred to by the tree entry with name "root"
    /// (it also has OID "root")
    /// (I put it here but it should be skipped and provided by a method)
    payload: Vec<u8>,
    /// Commit trailers
    signatures: HashMap<PublicKey, Signature>,
}
```

Struct `Doc` would disappear because its contents have been split between:

* the `Itentity::replaces` and `Identity::delegations` properties, and
* the application payload that is ignored by radicle link.

Depending on how we would handle merges in the workflow there could be a simpler encoding but this would be the basic step.

Reply:
Well, one issue I have with that is that not having a generic `payload` means we're constraining the type too much. We always have to interact with a `Vec<u8>`. If we pick a generic parameter, however, it gives us a degree of freedom for implementations and testing.

I'm also finding it hard to see your advantages here. In the above, all you've done is moved the `Doc` fields up. Thus, these are the same definitions (minus the generics).

Reply:
> We always have to interact with a `Vec<u8>`.

This is exactly the point of my proposal: not having a type at the radicle link level.

The point is who is the "we".
"We" as "radicle link" will just ignore that blob, we are is just meant to give Upstream (or whatever app layer we'll serve) access to the payload.

Then "we" as "Upstream" will deal with the blob, exactly as radicle link deals with other blobs at its own level.
Having envelops providing opaque payloads to upper layers is how most protocols are structured.

Reply:
One further explanation: soon we'll have to define how #212 works.
And there the idea is precisely to use some kind of envelop carrying opaque payloads.

There the need is obvious because we expect to have a lot of different event types: tying radicle link to those types, or even just making it aware of them, cannot possibly work.

What I am saying is that we should apply the same principle here.

Reply:
Apologies that I'm taking some time to see your vision :) I think I'm starting to see what you mean, however...

>  "We" as "radicle link" will just ignore that blob, we are is just meant to give Upstream (or whatever app layer we'll serve) access to the payload.

How and where does this happen? Link still has to store the blob of data and provide the hash for looking it up. Where does this go? 

> There the need is obvious because we expect to have a lot of different event types: tying radicle link to those types, or even just making it aware of them, cannot possibly work.

Can you explain how the current proposal is "tying radicle link to those types"? If the contents are generic in `T` we're not tying it to anything, at least that's my perspective.

Reply:
Massi, this is an unnecessary indirection, and your motivations don't justify it. Every application wishing to specialise `T` to custom payloads will have to deal with the fact that their payload may not actually be present, or that it was created by an earlier or later version of their software. We can make everyone's lives easier by using an extensible payload format (such as JSON-LD), which makes no guarantees, but preserves unknown schema parts in untyped form.